mongo/mongo.go

```diff
@@ -0,0 +1,362 @@
+func (db *DB) Get(bucket, key []byte) (ret []byte, err error) {
+	return db.get(bucket, key, context.Background())
+}
+
+// Set stores the given value on bucket and key.
+func (db *DB) Set(bucket, key, value []byte) error {
+	return db.set(bucket, key, value, context.Background())
+}
+
+// Del deletes the value stored in the given bucket and key.
+func (db *DB) Del(bucket, key []byte) error {
+	return db.del(bucket, key, context.Background())
+}
+
+func (db *DB) List(bucket []byte) ([]*database.Entry, error) {
+	return db.list(bucket, context.Background())
+}
+
+// CmpAndSwap modifies the value at the given bucket and key (to newValue)
+// only if the existing (current) value matches oldValue.
+func (db *DB) CmpAndSwap(bucket, key, oldValue, newValue []byte) ([]byte, bool, error) {
+	wc := writeconcern.Majority()
+	txnOptions := options.Transaction().SetWriteConcern(wc)
+
+	session, err := db.db.Client().StartSession()
+	if err != nil {
+		return oldValue, false, errors.Wrap(err, "error starting session")
+	}
+	defer session.EndSession(context.Background())
+
+	val, swapped := []byte{}, false
+	err = mongo.WithSession(context.Background(), session, func(ctx mongo.SessionContext) error {
+		if err = session.StartTransaction(txnOptions); err != nil {
+			return errors.Wrap(err, "error: pending transaction")
+		}
+
+		val, swapped, err = db.cmpAndSwap(bucket, key, oldValue, newValue, ctx)
+		if err != nil {
+			if err = session.AbortTransaction(context.Background()); err != nil {
+				return errors.Wrapf(err, "failed to execute CmpAndSwap transaction on %s/%s and failed to rollback transaction", bucket, key)
+			}
+			return errors.Wrap(err, "error aborting transaction")
+		}
+
+		if err = session.CommitTransaction(context.Background()); err != nil {
+			return errors.Wrap(err, "error committing transaction")
+		}
+		return nil
+	})
+
+	return val, swapped, err
+}
+
+// Update performs multiple commands on one read-write transaction.
+func (db *DB) Update(tx *database.Tx) error {
+	wc := writeconcern.Majority()
+	txnOptions := options.Transaction().SetWriteConcern(wc)
+
+	session, err := db.db.Client().StartSession()
+	if err != nil {
+		return errors.Wrap(err, "error starting session")
+	}
+	defer session.EndSession(context.Background())
+
+	err = mongo.WithSession(context.TODO(), session, func(ctx mongo.SessionContext) error {
+		if err = session.StartTransaction(txnOptions); err != nil {
+			return errors.Wrap(err, "error: pending transaction")
+		}
+
+		err = db.executeTransactions(tx, ctx, session)
+		if err != nil {
+			return err
+		}
+
+		if err = errors.WithStack(session.CommitTransaction(context.Background())); err != nil {
+			return errors.Wrap(err, "error committing transaction")
+		}
+		return nil
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// CreateTable creates a bucket or an embedded bucket if it does not exists.
+func (db *DB) createTable(bucket []byte, ctx context.Context) error {
```

Comment:
Whenever you provide a `ctx context.Context` as a function/method parameter, put it as the first one; not last.